R/method-correlation.R

```diff
@@ -0,0 +1,125 @@
+#' 
+#' @param data1 data.table with columns as variables. All columns must be numeric. One row per sample.
+#' @param data2 data.table with columns as variables. All columns must be numeric. One row per sample. Will correlate all columns of data2 with all columns of data1.
+#' @param method string defining the type of correlation to run. The currently supported values are 'spearman' and 'pearson'
+#' @param verbose boolean indicating if timed logging is desired
+#' @return ComputeResult object
+setMethod("correlation", signature("data.table", "data.table"), function(data1, data2, method = c('spearman','pearson'), verbose = c(TRUE, FALSE)) {
+
+  # Check that the number of rows match.
+  if (!identical(nrow(data1), nrow(data2))) {
+    stop("data1 and data2 must have the same number of rows.")
+  }
+
+  # Check that all values are numeric
+  if (!identical(veupathUtils::findNumericCols(data1), names(data1))) { stop("All columns in data1 must be numeric.")}
+  if (!identical(veupathUtils::findNumericCols(data2), names(data2))) { stop("All columns in data2 must be numeric.")}
+
+
+  ## Compute correlation
+  # Resulting data table has column "rn" = row names of the correlation matrix (so taxa names), and the other
+  # column names are the vars from sample metadata that we use
+  corrResult <- data.table::as.data.table(cor(data1, data2, method = method), keep.rownames = T)
+
+  veupathUtils::logWithTime(paste0('Completed correlation with method=', method,'. Formatting results.'), verbose)
+
+
+  ## Format results
+  meltedCorrResult <- melt(corrResult, id.vars=c('rn'))
+  formattedCorrResult <- data.frame(
+    data1 = meltedCorrResult[['rn']],
+    data2 = meltedCorrResult[['variable']],
+    correlationCoef = meltedCorrResult[['value']]
+  )
+
+  return(formattedCorrResult)
+
+})
+
+#' Correlation of abundance data and metadata
+#' 
+#' This function returns the correlation of all columns of an AbundanceData object with appropriate columns of a SampleMetadata object.
+#' 
+#' @param data1 AbundanceData object. Will correlate abundance variables with specified variables in data2
+#' @param data2 SampleMetadata object. Will assume that all non-ID columns of metadata are appropriate for correlation.
+#' @param method string defining the type of correlation to run. The currently supported values are 'spearman' and 'pearson'
+#' @param verbose boolean indicating if timed logging is desired
+#' @export
+setMethod("correlation", signature("AbundanceData", "SampleMetadata"), function(data1, data2, method = c('spearman','pearson'), verbose = c(TRUE, FALSE)) {
+
+  # Many of our checks betwee smaple metadata and data live in the object check function of the abudnance class. 
+  # So we can use those (runnning validObject) to do a bunch of checks, then access the data.
+  if (!is.null(data1@sampleMetadata)) {
+    warning("Replacing sampleMetadata in data1 with data2 sampleMetadata")
+  }
+  data1@sampleMetadata <- data2
```

Comment:
Using the SampleMetadata helps us be explicit about what type of correlation we're running. Let's say the correlation only used an AbundanceData object. Are we running taxa x taxa? Taxa x metadata? The metadata x metadata? This line feels odd i agree but it's the easiest way to run all of the appropriate checks on the metadata to make sure it works wiht the taxa abundances. I can make that more clear with some comments.

Reply:
I still don't like it.. what if the method was `correlateAbundancesAndMetadata` or something and took only the AbundanceData object? That's also clear and the API is cleaner. The metadata v metadata one could take a single object SampleMetadata as well no?

Reply:
I just feel like the generic data v data fxn is more an internal helper and these others should be methods of classes properly. And their API should make sense from that perspective and from the perspective of someone trying to use them rather than dev them.